service/dynamodb/expression/expression.go

```diff
@@ -0,0 +1,600 @@
+// //
+// // Example:
+// //
+// //     // let builder be an existing Builder{} and
+// //     // updateBuilder be an existing updateBuilder{}
+// //     builder = builder.WithUpdate(updateBuilder)
+// //
+// //     expression := builder.Build()
+// func (b Builder) WithUpdate(updateBuilder UpdateBuilder) Builder {
+// 	if b.expressionMap == nil {
+// 		b.expressionMap = map[expressionType]treeBuilder{}
+// 	}
+// 	b.expressionMap[update] = updateBuilder
+// 	return b
+// }
+//
+// // WithUpdate function will create a Builder with the argument
+// // updateBuilder as a child treeBuilder. Users will able to add other
+// // treeBuilders to the Builder or call Build() to build a Expression
+// // struct.
+// //
+// // Example:
+// //
+// //     // let updateBuilder and conditionBuilder be an existing
+// //     // UpdateBuilder and ConditionBuilder respectively.
+// //     builder := expression.WithUpdate(updateBuilder)
+// //
+// //     builder = builder.WithCondition(conditionBuilder)   // Adding a ConditionBuilder
+// //     expression := builder.Build()                      // Creating a Expression
+// func WithUpdate(updateBuilder UpdateBuilder) Builder {
+// 	ret := Builder{}
+// 	return ret.WithUpdate(updateBuilder)
+// }
+
+// Expression will be a struct that will be able to generate members to DynamoDB
+// inputs.
+// The idea of Builder and Expression is separated to be able to check the
+// format of the child treeBuilders and to return an error with the Build()
+// method for ExpressionBuilder.
+//
+// Example:
+//
+//     keyCond := expression.Key("someKey").Equal(expression.Value("someValue"))
+//     proj := expression.NamesList(expression.Name("aName"), expression.Name("anotherName"), expression.Name("oneOtherName"))
+//     builder := expression.WithKeyCondition(keyCond).WithProjection(proj)
+//     expression := builder.Build()
+//
+//     queryInput := dynamodb.QueryInput{
+//       KeyConditionExpression:    expression.KeyCondition(),
+//       ProjectionExpression:      expression.Projection(),
+//       ExpressionAttributeNames:  expression.Names(),
+//       ExpressionAttributeValues: expression.Values(),
+//       TableName: aws.String("SomeTable"),
+//     }
+type Expression struct {
+	expressionMap map[expressionType]treeBuilder
+}
+
+// treeBuilder interface will be fulfilled by builder structs that represent
+// different types of Expressions.
+type treeBuilder interface {
+	// buildTree will create the tree structure of exprNodes. The tree structure
+	// of exprNodes will be traversed in order to build the string representing
+	// different types of Expressions as well as the maps that represent
+	// ExpressionAttributeNames and ExpressionAttributeValues.
+	buildTree() (exprNode, error)
+}
+
+// Condition will return the *string corresponding to the Condition Expression
+// of the argument Expression. This method is used to satisfy the members of
+// DynamoDB input structs. If the argument Expression does not have a
```

Comment:
nit, `Expression` here generally isn't talked about as `argument`. In this context it is more similar to a type that has methods on it. e.g `If the Expression does not have a condition expression this method will return nil.`